File: go/0104E-二叉树的最大深度.go
/*
 * @Author: SukiEva
 * @Date: 2021-11-03 16:44:54
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */

package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// 递归
// 迭代直接层序遍历 + 深度就行
func maxDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}
	leftDepth := maxDepth(root.Left)
	rightDepth := maxDepth(root.Right)
	if leftDepth > rightDepth {
		return 1 + leftDepth
	} else {
		return 1 + rightDepth
	}
}

File: cpp/0104E-二叉树的最大深度.cpp
/*
 * @Author: SukiEva
 * @Date: 2021-12-21 10:05:25
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
#include <bits/stdc++.h>

#include "include/ListNode.h"
#include "include/TreeNode.h"

using namespace std;

class Solution {
   public:
    int depth = 0;

    void dfs(TreeNode* node, int cnt) {
        if (node == nullptr) {
            depth = max(depth, cnt);
            return;
        }
        dfs(node->left, cnt + 1);
        dfs(node->right, cnt + 1);
    }

    int maxDepth(TreeNode* root) {
        dfs(root, 0);
        return depth;
    }
};